### PR TITLE
Prevent casting to TraceProviderSdk

### PR DESCRIPTION
**Pull request by trask**
**Problem statement:**

If users take the static global providers that are returned from the API, and cast them to their SDK specific implementations, then those casts will fail under auto-instrumentation, because auto-instrumentation takes over the static global providers returned by the API and points them to it's embedded SDK.

**Proposal:**

Prevent users from doing this to get SDK-specific implementation, so that auto-instrumentation does not cause them ClassCastException unexpectedly.

The preferably way to get the SDK-specific implementation is through `OpenTelemetrySdk` anyways.

**Implementation:**

Wrap (obfuscate) the static global providers that are returned from the API to prevent users from casting them to their SDK specific implementation.

Unfortunately, `OpenTelemetrySdk` needs to access the `TracerSdkProvider` _somehow_, and with this change it can no longer perform the cast either, so there's a semi-obscured `unobfuscate()` method that the SDK can use to get access to the underlying `TracerSdkProvider`. If anyone has other ideas for how to hide this from users, but allow it from the SDK let me know!

**Terminology choice:**

An alternative naming scheme could be:

* `ObfuscatedTracerProvider` --> `WrappedTracerProvider`
* `Obfuscated` --> `Wrapped`
* `unobfuscate()` --> `unwrap()`

But I wanted to it be really clear to end users that this is not something they should do, and casting to something called `Wrapped` and calling `unwrap()` sounds much more likely for someone to do compared to casting to something called `Obfuscated` and calling `unobfuscate()`.

**Note:**

This PR only covers `TracerProvider`, but if/when there's agreement on this, we can extend to `MeterProvider` and `CorrelationContextManager`.

### CODE REVIEW

File: api/src/main/java/io/opentelemetry/internal/Obfuscated.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * <p>This is important for auto-instrumentation, because if users take the static global providers
+ * that are returned from the API, and cast them to their SDK specific implementations, then those
+ * casts will fail under auto-instrumentation, because auto-instrumentation takes over the static
+ * global providers returned by the API and points them to it's embedded SDK.
+ *
+ * @since 0.4.0
+ */
+@ThreadSafe
+public interface Obfuscated {
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/internal/Obfuscated.java:35]: Not sure I understand the purpose of this interface. Can we remove it?

File: api/src/main/java/io/opentelemetry/internal/Obfuscated.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  /**
+   * Returns the unobfuscated provider.
+   *
+   * @return the unobfuscated provider.
+   * @since 0.4.0
+   */
+  Object unobfuscate();
```
Comment (jkwatson) [inline api/src/main/java/io/opentelemetry/internal/Obfuscated.java:43]: Rather than `Object`, can we have this interface be parameterized with a type?

**Comment by jkwatson** [comment]
Is the idea here, that the `Obfuscated` interface is a hook for the auto-instrumentation to work with, by instrumenting the `unobfuscate` method and returning the bare SDK implementation? Or something else, entirely. 

In short, color me confused about how this "prevents casting". Maybe a little code snippet that demonstrates how it does the prevention would be helpful.

File: api/src/main/java/io/opentelemetry/internal/Obfuscated.java (hunk 2 above)
Comment (trask) [reply api/src/main/java/io/opentelemetry/internal/Obfuscated.java]: I like that idea. Done.

**Comment by carlosalberto** [comment]
I can see the reasons behind this (with which I agree with), but I'm afraid that 'wrapping' the `TracerProvider` for `OpenTelemetry` **and not** for `OpenTelemetrySdk` will cause headaches to users eventually.

File: api/src/main/java/io/opentelemetry/internal/Obfuscated.java (hunk 2 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/internal/Obfuscated.java]: @trask before doing this change, can we clarify why having this interface at all is necessary?

**Comment by trask** [comment]
@jkwatson Here's a code snippet, does this help explain better?

So currently, this works:

```
TracerSdkProvider tracerProvider = (TracerSdkProvider) OpenTelemetry.getTracerProvider();
```

and this PR prevents it from working, as the above code will now throw ClassCastException:

```
java.lang.ClassCastException: io.opentelemetry.OpenTelemetry$ObfuscatedTracerProvider
cannot be cast to io.opentelemetry.sdk.trace.TracerSdkProvider
```

which will make sure that users always use `OpenTelemetrySdk` to get the `TracerSdkProvider`:

```
TracerSdkProvider tracerProvider = OpenTelemetrySdk.getTracerProvider();
```

File: sdk/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java [hunk 3]
```diff
@@ -40,7 +41,7 @@
    * @since 0.1.0
    */
   public static TracerSdkProvider getTracerProvider() {
-    return (TracerSdkProvider) OpenTelemetry.getTracerProvider();
+    return (TracerSdkProvider) ((Obfuscated<?>) OpenTelemetry.getTracerProvider()).unobfuscate();
```
Comment (trask) [inline sdk/src/main/java/io/opentelemetry/sdk/OpenTelemetrySdk.java:44]: @bogdandrutu this is where the `Obfuscated` interface is used. `OpenTelemetrySdk.getTracerProvider()` needs some way to unwrap and retrieve the `TracerSdkProvider`. Did you have other idea how to unwrap?

**Comment by jkwatson** [comment]
> @jkwatson Here's a code snippet, does this help explain better?
> 
> So currently, this works:
> 
> ```
> TracerSdkProvider tracerProvider = (TracerSdkProvider) OpenTelemetry.getTracerProvider();
> ```
> 
> and this PR prevents it from working, as the above code will now throw ClassCastException:
> 
> ```
> java.lang.ClassCastException: io.opentelemetry.OpenTelemetry$ObfuscatedTracerProvider
> cannot be cast to io.opentelemetry.sdk.trace.TracerSdkProvider
> ```
> 
> which will make sure that users always use `OpenTelemetrySdk` to get the `TracerSdkProvider`:
> 
> ```
> TracerSdkProvider tracerProvider = OpenTelemetrySdk.getTracerProvider();
> ```

oh, ok. It makes it fail always at runtime, sure. I thought the goal was to making casting not even compile, so people couldn't even try it. gotcha

File: api/src/main/java/io/opentelemetry/internal/Obfuscated.java (hunk 2 above)
Comment (trask) [reply api/src/main/java/io/opentelemetry/internal/Obfuscated.java]: @bogdandrutu just wanted to make sure that you saw my response https://github.com/open-telemetry/opentelemetry-java/pull/1152#discussion_r416841711 (I should have replied here also)

**Pull request merged**